[3/n][Optimus][Auto-AC] Support float8_e4m3fn quantization type and set scaling as the default

mengluy0125 opened the pull request:
Summary:
1. Customers now can test with float8_e4m3fn.
2. To play safe, we set the scaling version as the default.

Test Plan:
### unit test
```
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:quantization
```

Buck UI: https://www.internalfb.com/buck2/f679f362-8bf4-454c-87df-a85cbc2ab2a8
Test UI: https://www.internalfb.com/intern/testinfra/testrun/5066549861047443
Network: Up: 16KiB  Down: 3.9MiB  (reSessionID-98badbfd-76f7-487f-ab1c-1ec4f850614d)
Analyzing targets. Remaining     0/281                                                                                                    
Executing actions. Remaining     0/5957                                                                                                   7.3s exec time total
Command: test.     Finished 3 local, 1 remote                                                                                                                                                                  
Time elapsed: 1:29.7s
Tests finished: Pass 3. Fail 0. Fatal 0. Skip 0. Build failure 0

Differential Revision: D74910193




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -517,6 +517,9 @@ def calculate_range(dtype: torch.dtype) -> tuple:
         max_val = 57344.0
+    elif dtype == torch.float8_e4m3fn:
+        min_val = -448
+        max_val = 448
```
@drisspg do we have any more programatic way of getting these values? (especially given that these are not the only two float8 types)

drisspg replied:
```Py
>>> torch.finfo(torch.float8_e4m3fn)
finfo(resolution=1, min=-448, max=448, eps=0.125, smallest_normal=0.015625, tiny=0.015625, dtype=float8_e4m3fn)
```

mengluy0125 replied:
Thanks for the context. Will update this.

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -601,6 +604,7 @@ def quantize_activation_bw(graph: torch.fx.Graph) -> None:
                         for bwd_input in bw_inputs
                         if bwd_input.name == scale_name
                     )
+                with graph.inserting_after(scale_node):
```
what made this wrong before? (was this a silent correctness issue?)

mengluy0125 replied:
Yes. We found a dynamo error when we run on LWB omni-FM. From tlparse (https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-v2_t1_e4m3_300-a5ff0a2445/attempt_0/version_0/rank_0/-_30_1_0/dynamo_error_1662.txt?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000), I found that if the order of the nodes are not topologically sorted, we will have the keyword error.

bdhirsh replied:
does the test cover this?

eellison replied:
Would it be possible to use `match.replace_by_example` or other pattern matching facilities? this is more robust, easier to follow, less bug prone..

bdhirsh commented on test/inductor/test_quantization.py:
```diff
@@ -159,6 +175,59 @@ def test_activation_quantization_aten(self):
+        post_grad_fusion_options={
+            "activation_quantization_aten_pass": {
+                "quant_type": "torch.float8_e4m3fn",
+                "use_scaling": False,
```
is there a benefit in `use_scaling=False`? (if so, it would be nice to clean up the extra configs / remove the extra code paths if we know we don't need them)

mengluy0125 replied:
use_scaling=False potentially will have better QPS and more memory saving. In our experiments on LWB omniFM, without scaling we can have 11% peak memory reduction, but with scaling, we only have 7%.

mengluy0125 committed:
[3/n][Optimus][Auto-AC] Support float8_e4m3fn quantization type and set scaling as the default (#153802)

Summary:

1. Customers now can test with all kinds of quant dtype.
2. To play safe, we set the scaling version as the default.
3. rename the scale_ to fp8_scale, and quant_ to fp8_quant_ since we obsereve duplicate argument ([https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/aps-v2_1024_scale_opt-ce794ca509/attempt_0/version_0/rank_0/-_35_0_0/dynamo_error_1076.txt?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000])

Test Plan:
### unit test
```
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/inductor:quantization
```

Buck UI: https://www.internalfb.com/buck2/f679f362-8bf4-454c-87df-a85cbc2ab2a8
Test UI: https://www.internalfb.com/intern/testinfra/testrun/5066549861047443
Network: Up: 16KiB  Down: 3.9MiB  (reSessionID-98badbfd-76f7-487f-ab1c-1ec4f850614d)
Analyzing targets. Remaining     0/281
Executing actions. Remaining     0/5957                                                                                                   7.3s exec time total
Command: test.     Finished 3 local, 1 remote
Time elapsed: 1:29.7s
Tests finished: Pass 3. Fail 0. Fatal 0. Skip 0. Build failure 0


#### E2E

thanks to Hahu803
N7138601

Reviewed By: Hahu803

Differential Revision: D74910193

jeffdaily commented:
I believe the problem is that f8 type e4m3fn is not supported on MI200 which is one of our CI servers.  The MI300 tests seem to be passing even though the MI300 HW uses e4m3fnuz.  You'll need to conditionally skip this failing test for MI200 only.  Example.

```python
from torch.testing._internal.common_utils import MI300_ARCH, skipIfRocmArch
@skipIfRocmArch(MI300_ARCH)
def test_the_thing(self):
   pass
```
